File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector < string > res;
        string s = "";
        solve( 0 , n , res , n , n , s);
        return res;
    }
    
    void solve( int x , int len , vector<string>& res , int open , int close , string s){
        if( x == 2 * len ){
            res.push_back(s);
            return;
        }
        
        if( open > 0 && close >= open)
            solve( x + 1 , len , res , open - 1 , close , s + '(');
        if( close > 0 && open <= close)
            solve( x + 1 , len , res , open , close - 1 , s + ')');
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.swift
class Solution {
    func generateParenthesis(_ n: Int) -> [String] {
        var result = [String]()
        backtrack("",0,0,n){
            result.append($0)
        }
        return result
    }
    
    func backtrack( _ current : String, _ open: Int, _ end: Int , _ max : Int, _ result: (String) -> Void ){
        if current.count == max * 2 {
            result(current)
            return
        }
        
        if open < max { backtrack(current + "(" , open + 1 , end , max , result )}
        if end < open { backtrack(current + ")" , open , end + 1 , max , result )}
    }
}